Fix raw_post raising when rack.input is nil

skipkayhil opened the pull request:
### Motivation / Background

Fixes #52125

### Detail

Starting in Rack 3.1, rack.input is optional, so `read_body_stream` (used by `raw_post`) can no longer call `read` unconditionally.

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Unrelated changes should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [x] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

skipkayhil committed:
Fix raw_post raising when rack.input is nil

Starting in Rack 3.1, rack.input is optional, so `read_body_stream`
(used by `raw_post`) can no longer call `read` unconditionally.

zzak commented on actionpack/lib/action_dispatch/http/request.rb:
```diff
@@ -466,11 +466,13 @@ def default_session
       end
 
       def read_body_stream
-        reset_stream(body_stream) do
```
Is there anything else that calls `reset_stream`? It is private, so maybe we should move this into there to deal with nil?

rafaelfranca replied:
Yes, you can move there. I actually refactored that code to fix this issue. I'm not sure only changing this method is enough. Rails has assumption that `request.body` will return something.

<img width="573" alt="Screenshot 2024-06-17 at 2 49 30 PM" src="https://github.com/rails/rails/assets/47848/2e30897f-a486-4e0f-980c-0319480ad0fb">

skipkayhil replied:
I felt like `read_body_stream` was an appropriate place to check for `nil` since `reset_stream` to me implies we are passing a (non-nil) `stream`, but I don't have a strong preference and am happy to move the nil-check.

Is the goal to have `body` always return a `StringIO` even though `rack.input` can be `nil` now? Or just that we need to handle that case in Action Mailbox?

rafaelfranca replied:
I think `body` can return `nil`, but our code should not rely on it never returning `nil`.

skipkayhil committed:
Fix Action Mailbox assuming request.body present

In Rack 3.1, `rack.intput` (`request.body`) is no longer guaranteed to
be present, so we can no longer unconditionally call `read` on it.

Co-authored-by: zzak <zzakscott@gmail.com>

skipkayhil commented:
Rebased and fixed the Action Mailbox controller. As I mention in the commit message, this doesn't appear to be testable inside `IntegrationTest` at the moment. It seems like `rack` and/or `rack-test` needs to be updated to allow `rack.input` to be `nil` inside `IntegrationTest`.

zzak commented:
Maybe this helps:

* `Rack::MockRequest.env_for` accepts an input option:
https://github.com/rack/rack/blob/0b6f760021e01b3031128e81536fc61cef676969/lib/rack/mock_request.rb#L142
* You can also pass in options to `MockRequest#request`: https://github.com/rack/rack/blob/0b6f760021e01b3031128e81536fc61cef676969/lib/rack/mock_request.rb#L66-L67
* We configure the MockRequest in rails here: https://github.com/rails/rails/blob/6045f257fbbff039edb2542c9859488cec0dcaa9/actionpack/lib/action_dispatch/testing/test_request.rb#L17-L21

So maybe you can pass it in your test? Or you can override `env[RACK_INPUT]`?

I'm not sure it helps, but it's nice if we have test coverage, I guess.

Thanks for working on this!

skipkayhil commented:
> Or you can override `env[RACK_INPUT]`

This worked 🎉 I rebased to fix the CHANGELOG conflict and now there is a working test for both places

jimbali commented:
Wow, thanks for fixing and releasing this one day before I ran into the issue! 🎉
